src/oop/user_defined_classes/hard/q2/MainTest.java

```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Test
+    public void dishWasherClassHasRequiredConstructor() {
+        String missingConstructorMessage = "Your " + testClassName + " class is missing a required constructor.";
+        Class<?>[] arguments= {
+                int.class, int.class
+        };
+        assertTrue(classInstance.hasConstructor(arguments), missingConstructorMessage);
+    }
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of(0, 0),
+                Arguments.of(35, 12),
+                Arguments.of(478273, 723662367)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void dishWasherConstructorInitializesFieldsCorrectly(int attribute1, int attribute2) throws Throwable{
+        Object[][] arguments = {
+                {attribute1, int.class},
+                {attribute2, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        String incorrectFieldInstantiationMessage =
+                "Your " + testClassName + " constructor does not correctly initialize the object." ;
+        _assertEquals(attribute1, classInstance.getFieldValue(instance, intAttributeName1),
+                incorrectFieldInstantiationMessage);
+        _assertEquals(attribute2, classInstance.getFieldValue(instance, intAttributeName2),
+                incorrectFieldInstantiationMessage);
+    }
+
+    private static Stream<Arguments> methodInputProvider() {
+        return Stream.of(
+                Arguments.of(2, 3, 1, 1, 1, 2),
+                Arguments.of(15, 33, 14, 15, 1, 18),
+                Arguments.of(378, 32891, 378, 32891, 0, 0),
+                Arguments.of(0, 0, 0, 0, 0, 0),
+                Arguments.of(5, 10, 0, 1, 5, 9)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctLoadDishWasherMethod(int attribute1, int attribute2, int remove1, int remove2, int remainder1, int remainder2) throws Throwable{
+        Object[][] arguments = {
+                {attribute1, int.class},
+                {attribute2, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        assertTrue(classInstance.hasMethod(methodName, new Class[]{int.class, int.class}),
+                "Your " + testClassName + " class is missing the method " + methodName + ".");
+        classInstance.callMethod(
+                methodName, new Object[][]{{remove1, int.class}, {remove2, int.class}}, null, instance, new Clause[]{
+                        new IntegerLiteral(remainder1),
+                        new StringLiteral(" " + intAttributeName1 + " are still clean"),
+                        new NewLine(),
+                        new IntegerLiteral(remainder2),
+                        new StringLiteral(" " + intAttributeName2 + " are still clean"),
+                        new NewLine()
```

Comment:
I think its better to add optional space at the end of the outputs

Reply:
I checked and it looks like the clause testing adds an optional newLine at the end anyways so I just got rid of my new line and let that invisible one take its place